Hadrons/DistilMatrix.hpp

```diff
@@ -26,89 +26,289 @@
+
+    void saveBlock(const DistilMatrixSet<T> &m, const unsigned int ext, const unsigned int str,
+                               const unsigned int i, const unsigned int j, std::string datasetName,
+                               const unsigned int t, const unsigned int chunkSize);
+
+    template <template <class> class Vec, typename VecT>
+    void load(Vec<VecT> &v, const unsigned int t, const std::string dataset_name, double *tRead = nullptr, GridBase *grid = nullptr);
+private:
+    std::string  filename_{""}, dataname_{""};
+    unsigned int nt_{0}, ni_{0}, nj_{0};
+};
+
+// implementation /////////////////////////////////////////////////////////////////
+template <typename T>
+DistilMatrixIo<T>::DistilMatrixIo(std::string filename, std::string dataname, 
+                            const unsigned int nt, const unsigned int ni,
+                            const unsigned int nj)
+: filename_(filename), dataname_(dataname), ni_(ni), nj_(nj), nt_(nt)
+{}
+
+template <typename T>
+template <typename MetadataType>
+void DistilMatrixIo<T>::initFile(const MetadataType &d)
+{
 #ifdef HAVE_HDF5
-        // variable-length struct
-        typedef struct  {
-            size_t len; /* Length of VL data (in base type units) */      
-            void *p;    /* Pointer to VL data */        
-        } VlStorage;
-        
-        Hdf5Reader  reader(fileName_, false);
-        push(reader, metadataName_);
-        H5NS::Group &subgroup = reader.getGroup();
-
-        if(!newgroupname.empty())
-        {
-            H5NS::Exception::dontPrint();
-            try{
-                subgroup = subgroup.openGroup(newgroupname);
-            } catch (...) {
-                subgroup = subgroup.createGroup(newgroupname);
-            }
-        }
+    Hdf5Writer writer(filename_);
+    push(writer, dataname_);    //creates main h5 group
+    write(writer, "Metadata", d);
+#else
+    HADRONS_ERROR(Implementation, "distil matrix I/O needs HDF5 library");
+#endif
+}
+
+template <typename T>
+size_t DistilMatrixIo<T>::getSize(void) const
+{
+    return nt_*ni_*nj_*sizeof(T);
+}
+
+template <typename T>
+void DistilMatrixIo<T>::saveBlock(const T *data, 
+                               const unsigned int i, 
+                               const unsigned int j,
+                               const unsigned int blockSizei,
+                               const unsigned int blockSizej,
+                               std::string t_name,
+                               std::string datasetName,
+                               const unsigned int chunkSize)
+{
+#ifdef HAVE_HDF5
+    
+    H5NS::H5File file(filename_,H5F_ACC_RDWR);
+    H5NS::Group rootgroup = file.openGroup(DISTIL_MATRIX_NAME);
+
+    // create t group if it doesnt exist, or open it
+    H5NS::Group tgroup;
+    // if ( H5Lexists( file.getId(), t_name, H5P_DEFAULT ) > 0 )
```

Comment:
why these commented out bits? do we need them later or can we delete them?

Reply:
Ok, this is something I couldn't solve yet. Instead of using the try catch block below it (which apparently works but Antonin said it's not ideal), I should use the code commented out here. But it's the C api for hdf5, and for some reason I couldn't make it see the function H5Lexists